develop-csharp-plugins.md

```diff
@@ -1,7 +1,71 @@
+Flow is written in C#, so plugins written in dotnet platform can directly communicate with Flow without extra protocols.
 
-For now you can use the [dotnet template](https://github.com/Flow-Launcher/dotnet-template) to generate your plugin 
+## Initialization
+
+We recommand you use the [dotnet template](https://github.com/Flow-Launcher/dotnet-template) to generate a plugin template first.
+
+In order to be recongized as a Flow DotNet plugin, the directory need to have at least two files
+1. [`plugin.json`](plugin.json.md)
+2. A Dotnet class that implement **[IPlugin](API-Reference/Flow.Launcher.Plugin/iplugin.md)** or **[IAsyncPlugin](API-Reference/Flow.Launcher.Plugin/iasyncplugin.md)** (remember to refrence [Flow.Launcher.Plugin](https://www.nuget.org/packages/Flow.Launcher.Plugin/) by Nuget). The plugin template will add the reference and create  a`Main.cs` that implement `IPlugin`.
+
+Found our API Reference [here](API-Reference/)
+
+
+A sample CSharp Plugin [here](https://github.com/Flow-Launcher/plugin-samples/tree/master/HelloWorldCSharp) 
+
+## IPlugin/IAsyncPlugin
+
+The `Main`class that implement **[IPlugin](API-Reference/Flow.Launcher.Plugin/iplugin.md)** or **[IAsyncPlugin](API-Reference/Flow.Launcher.Plugin/iasyncplugin.md)** will handle the query search with Flow.
+
+**[IPlugin](API-Reference/Flow.Launcher.Plugin/iplugin.md)** interface contains two required method
+1. `void Init(PluginInitContext context)`
+    - [PluginInitContext](https://github.com/Flow-Launcher/Flow.Launcher/blob/master/API-Reference/Flow.Launcher.Plugin/PluginInitContext.cs) exposes some API from Flow and an metadata object for your plugin. 
+    - `Init`method will be invoked before the invocation of `Query`, so you can do some preparation here. 
+    - We recommand you do expensive operations in `Init`instead of Object Constructor because `Init`method will be executed in parallel with other plugin.
+2. `List<Result> Query(Query query)`
+    - `Query`will be invoked when user activate this plugin with specific ActionKeyword.
+    - A `List`of [Result](/API-Reference/Flow.Launcher.Plugin/result.md) object should be returned.
+ 
+ **[IAsyncPlugin](API-Reference/Flow.Launcher.Plugin/iasyncplugin.md)** is the async version of **[IPlugin](API-Reference/Flow.Launcher.Plugin/iplugin.md)**
+ - Instead of implmenting `Init` and `Query`, you will need to implement `InitAsync`and `QueryAsync`, which use `Task`,`Task<List<Result>` as return value to allow  using `async/await`strategy
+ - `QueryAsync` provides a `CancellationToken token` to allow you to check whether user has typed a new query.
+
+
+## Additional Interface
+
+Besides the basic implementation of **IPlugin/IAsyncPlugin**, plugins can also implement a series of interface that belongs to **IFeatures** to control more communication with Flow. 
+
+**Remarks**: We requires you to implement these interfaces in the same class that implement **IPlugin/IAsyncPlugin**.
```

Comment:
"We requires you to" -> "You should"
"same class that implement" -> "same class that implements"